gio/src/unix_fd_list.rs

```diff
@@ -47,41 +47,107 @@ pub trait UnixFDListExtManual: sealed::Sealed + IsA<UnixFDList> + Sized {
             }
         }
     }
 
     #[doc(alias = "g_unix_fd_list_peek_fds")]
-
-    fn peek_fds(&self) -> Vec<RawFd> {
+    fn peek_fds(&self) -> &[BorrowedFd<'_>] {
         unsafe {
             let mut length = mem::MaybeUninit::uninit();
-            let ret = FromGlibContainer::from_glib_none_num(
-                ffi::g_unix_fd_list_peek_fds(self.as_ref().to_glib_none().0, length.as_mut_ptr()),
+            std::slice::from_raw_parts(
+                ffi::g_unix_fd_list_peek_fds(self.as_ref().to_glib_none().0, length.as_mut_ptr())
+                    as *const BorrowedFd,
                 length.assume_init() as usize,
-            );
-            ret
+            )
         }
     }
+
     #[doc(alias = "g_unix_fd_list_steal_fds")]
-    fn steal_fds(&self) -> Vec<RawFd> {
+    fn steal_fds(&self) -> FdArray {
         unsafe {
             let mut length = mem::MaybeUninit::uninit();
-            let ret = FromGlibContainer::from_glib_full_num(
-                ffi::g_unix_fd_list_steal_fds(self.as_ref().to_glib_none().0, length.as_mut_ptr()),
-                length.assume_init() as usize,
-            );
-            ret
+
+            let ptr =
+                ffi::g_unix_fd_list_steal_fds(self.as_ref().to_glib_none().0, length.as_mut_ptr());
+
+            FdArray {
+                ptr: ptr::NonNull::new(ptr).unwrap(),
+                len: length.assume_init() as usize,
+            }
         }
     }
 }
 
 impl<O: IsA<UnixFDList>> UnixFDListExtManual for O {}
+
+pub struct FdArray {
+    ptr: ptr::NonNull<libc::c_int>,
+    len: usize,
+}
+
+pub struct FdIterator {
+    ptr: ptr::NonNull<libc::c_int>,
+    len: usize,
+}
+
+impl Iterator for FdIterator {
+    type Item = OwnedFd;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.len > 0 {
+            let current = self.ptr.as_ptr();
+            if self.len > 1 {
+                let next = unsafe { self.ptr.as_ptr().add(1) };
+                self.ptr = ptr::NonNull::new(next).unwrap();
+            }
+            self.len -= 1;
+            Some(unsafe { OwnedFd::from_raw_fd(*current) })
+        } else {
+            None
+        }
+    }
+}
+
+impl Drop for FdArray {
+    fn drop(&mut self) {
+        while self.len > 0 {
+            unsafe {
+                let current = self.ptr.as_ptr();
+                libc::close(*current);
+            }
+            if self.len > 1 {
+                let next = unsafe { self.ptr.as_ptr().add(1) };
+                self.ptr = ptr::NonNull::new(next).unwrap();
+            }
+            self.len -= 1;
+        }
+    }
+}
+
+impl std::iter::IntoIterator for FdArray {
+    type Item = OwnedFd;
+    type IntoIter = FdIterator;
+
+    fn into_iter(mut self) -> Self::IntoIter {
+        // Can I .take() the ptr and set it to null here?
+        let len = std::mem::take(&mut self.len);
+        FdIterator { len, ptr: self.ptr }
+    }
+}
+
+impl FdArray {
+    pub fn as_slice(&self) -> &[BorrowedFd<'_>] {
```

Comment:
As you didn't implement `Deref<Target = [OwnedFd]>`, it would be nice if you implemented a non-destructive `iter()` too. Personally, I would just implement `Deref` :)